kong/clustering.lua

```diff
@@ -204,38 +206,58 @@ local function communicate(premature, conf)
     end
   end)
 
-  while not exiting() do
-    local data, typ, err = c:recv_frame()
-    if err then
-      ngx.log(ngx.ERR, "error while receiving frame from control plane: ", err)
-      c:close()
+  local main_thread = ngx.thread.spawn(function()
+    while not exiting() do
+      local data, typ, err = c:recv_frame()
+      if err then
+        if not is_timeout(err) then
```

Comment:
So, for the DP side of code, when `c:recv_frame` timedous, we should also be checking to see if this has been overdue compared to `PING_WAIT`, just like the CP did. These code can largely be duplicated from the CP side.